rust-spec/lightclient/detection/detection.md

```diff
@@ -276,421 +443,356 @@ This implies that we need a timeout of at least *2 Delta* for remote
+## Outline
 
-- Error condition
-    - if precondition is violated
+The problem laid out is solved by calling the function `AttackDetector`
+with a lightstore that contains a light block that has just been
+verified by the verifier.
 
-#### **[LCD-FUNC-REPLACE-SECONDARY.1]:**
+Then `AttackDetector` downloads headers from the secondaries. In case
+a conflicting header is downloaded from a secondary,
+`CreateEvidenceForPeer` which computes evidence in the case that
+indeed an attack is confirmed. It could be that the secondary reports
+a bogus block, which means that there need not be an attack, and the
+secondary is replaced.
+  
 
-```go
-Replace_Secondary(addr Address)
-```
+  
+## Details of the functions
 
-**TODO:** formalize conditions
+#### **[LCD-FUNC-DETECTOR.1]:**
 
-- Implementation remark
-    - maintain [LCD-INV-TRUSTED-AGREED.1], that is,
-        - call `CrossCheck(nsec,lightStore.LatestTrusted()`.
-           If it matches we are OK, otherwise
-            - we might just repeat with another full node as new secondary
-            - **FUTURE:** try to do fork detection from some possibly old
-             lightblock in store. (Might be the approach for the
-             light node that assumes to be connected to correct
-             full nodes only from time to time)
+```go
+func AttackDetector(root LightBlock, primary_trace []LightBlock) 
+                   ([]InternalEvidence) {
+   
+    Evidences := new []InternalEvidence;
+ 
+    for each secondary in Secondaries {
+        // we replay the primary trace with the secondary, in
+        // order to generate evidence that we can submit to the
+        // secodary. We return the evidence + the trace the
+        // secondary told us that spans the evidence at its local store
+   
+        EvidenceForSecondary, newroot, secondary_trace, result :=
+                CreateEvidenceForPeer(secondary, 
+				                      root, 
+									  primary_trace);
+        if result == FaultyPeer {
+            Replace_Secondary(root);
+        }
+        else if result == FoundEvidence {
+            // the conflict is not bogus
+            Evidences.Add(EvidenceForSecondary);
+            // we replay the secondary trace with the primary, ...
+            EvidenceForPrimary, _, result := 
+				    CreateEvidenceForPeer(primary,
+                                          newroot,
+                                          secondary_trace);
+            if result == FoundEvidence {
+                Evidences.Add(EvidenceForPrimary);
+            }
+            // At this point we do not care about the other error
+            // codes. We already have generated evidence for an
+            // attack and need to stop the lightclient. It does not
+            // help to call replace_primary. Also we will use the
+            // same primary to check with other secondaries in
+            // later iterations of the loop
+         }
+         else {
```

Comment:
I don't get what this branch is doing.

Reply:
Sorry. I changed the control flow last second. I will fix that right now. We do nothing here. This is the case where the secondary agrees with the primary.

Reply:
fixed